### PR TITLE
Set Open_vSwitch hardware offload in config daemon

### PR DESCRIPTION
**Pull request by SalDaniele**
Previously, we have set ovs hardware offload via the service applied by the Machine Config Operator
(bindata/manifests/switchdev-config/ovs-units.ovs-vswitchd.service).

However, the issue with this approach is that after the user has removed the MachineConfigPool or removed SRIOV all together, ovs hw-offload remains enabled.

This change will instead handle configuring ovs hardware offloading in the config daemon by checking for switchdev devices, and enabling ovs hardware offload only while such devices are detected. We will remove the serviced file responsible for setting ovs-hwol to true on boot. Upon detecting a change in the sriov node policy, we will check for any switchdev devices and set ovs-hwol accordingly. Changes to ovs will result in drain and reboot of the node to ensure changes are applied

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -491,6 +491,15 @@ func (dn *Daemon) nodeStateSyncHandler() error {
 			}
 		}
 	}
+
+	d, r, err := trySetOvsHWOL(utils.SwitchdevDeviceExists(latestState))
```
Comment (bn222) [inline pkg/daemon/daemon.go:495]: iiuc, we are adding 1 more reboot when we enable switchdev hw offload since that code path also uses this and in that case, we already have to have an early dev-link command on each boot. Is my understanding correct? In other words, we pay an extra reboot at the time we enable it to gain a way to get rid of it when we disable it. 

Thinking along the same lines, would it even make sense to have ovs set dev-link to switchdev mode? I'm not sure we want that, but then we could go back to one reboot.

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -1079,3 +1088,42 @@ func tryCreateNMUdevRule() error {
...
+	// if we are already in desired state, do nothing
+	cmd := exec.Command("ovs-vsctl", "get", "Open_vSwitch", ".", "other_config:hw-offload")
+	cmd.Stdout = &output
+	if err := cmd.Run(); err != nil {
+		// if ovs has not previously been configured this record may not yet exist
+		glog.Errorf("trySetOvsHWOL(): failed to get current Open_vSwitch hw-offload mode : %v", err)
+		currentState = "false"
+	} else {
+		currentState = output.String()
+	}
```
[the remark this answers is not in the document]
Comment (bn222) [reply pkg/daemon/daemon.go:1114]: Would also mean we don't need to convert the bool into a string. It makes a bit more sense to compare bool with bool instead of string that should look like a bool with a string that was created from a bool.

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -1079,3 +1088,42 @@ func tryCreateNMUdevRule() error {
...
+
+	if strings.Contains(currentState, desiredState) {
+		glog.V(2).Infof("trySetOvsHWOL(): ovs already in desired state %s, ovs switch not needed", desiredState)
+		return false, false, nil
+	}
+
+	// otherwise switch state and reboot to apply the config changes to ovs
+	glog.V(2).Infof("trySetOvsHWOL(): switching Open_vSwitch hw-offload from %s to %s", currentState, desiredState)
+	cmd = exec.Command("ovs-vsctl", "set", "Open_vSwitch", ".", fmt.Sprintf("other_config:hw-offload=%s", desiredState))
```
Comment (bn222) [inline pkg/daemon/daemon.go:1123]: I'm starting to think we should have a function that wraps `exec.Command("ovs-vsctl", ...)`. I leave it up to you if you think we benefit from not having to copy paste `exec.Command("ovs-vsctl",` every time we want to call ovs-vsctl

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -1079,3 +1088,42 @@ func tryCreateNMUdevRule() error {
 	}
 	return nil
 }
+
+func trySetOvsHWOL(newState bool) (bool, bool, error) {
```
Comment (bn222) [inline pkg/daemon/daemon.go:1092]: newState is too vague. Can we come up with a bbetter name?

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (wizhaoredhat) [reply pkg/daemon/daemon.go:495]: In that note. If we make the hw-offload config in OVS ephemeral then we can avoid this all together. Having an ephemeral config has it's pros and cons.

Comment (SalDaniele) [reply pkg/daemon/daemon.go:495]: Yes, we are currently paying an extra reboot when the daemon detects a device has been put into switchdev mode.

Comment (wizhaoredhat) [reply pkg/daemon/daemon.go:495]: One potential issue is that pods might become Ready with a VF while HWOL is turned off. I don't think there will be network connectivity for those pods.

File: pkg/daemon/daemon.go (hunk 4 above)
Comment (SalDaniele) [reply pkg/daemon/daemon.go]: Updated to "switchdevDevExists"

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (SalDaniele) [reply pkg/daemon/daemon.go]: I agree, added functions to wrap running ovs-vsctl and restarting ovs.

File: pkg/daemon/daemon.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SalDaniele) [reply pkg/daemon/daemon.go]: Added "func OvsHWOLisEnabled()"

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (SalDaniele) [reply pkg/daemon/daemon.go]: See the most recent update, we can instead manually restart ovs from the config daemon to avoid the extra restart.

With the new behavior, we will trigger configuring ovs-hwol after a node reboots due to a change in the node state policy. We will detect if the node has booted with or without switchdev devices and set ovs-hwol accordingly

Comment (bn222) [reply pkg/daemon/daemon.go]: I'm ok with a restart of ovs. In theory, it should be possible to dynamically take care of this as well to be honest.

File: pkg/daemon/daemon.go [hunk 5]
```diff
@@ -491,6 +491,7 @@ func (dn *Daemon) nodeStateSyncHandler() error {
 			}
 		}
 	}
+
```
Comment (e0ne) [inline pkg/daemon/daemon.go:494]: this change is not related

File: pkg/daemon/daemon.go [hunk 6]
```diff
@@ -1079,3 +1085,38 @@ func tryCreateNMUdevRule() error {
...
+	desiredHWOLStateStr := strconv.FormatBool(switchdevDevExists)
+
+	exit, err := utils.Chroot("/host")
+	if err != nil {
+		glog.Errorf("trySetOvsHWOL(): failed to chroot to host: %v", err)
+	} else {
+		defer exit()
+	}
+
+	currentState := utils.OvsHWOLisEnabled()
```
Comment (e0ne) [inline pkg/daemon/daemon.go:1100]: please rename currentState -> hwOffloadEnabled

File: pkg/utils/utils.go [hunk 7]
```diff
@@ -788,6 +788,22 @@ func RunCommand(command string, args ...string) (string, error) {
...
+	cmd := exec.Command("ovs-vsctl", args...)
+	cmd.Stdout = &output
+	err := cmd.Run()
+
+	return output.String(), err
+}
+
+func RestartOVS() error {
+	glog.Infof("RestartOVS(): restarting openvswitch.service")
+	cmd := exec.Command("systemctl", "restart", "openvswitch.service")
```
Comment (e0ne) [inline pkg/utils/utils.go:803]: did you test it ith Ubuntu 22.04? AFAIK, service name in Ubuntu is different

**Review by e0ne — CHANGES_REQUESTED** [review]
That's good improvement but the PR needs to be updated. Please, add unit-tests too

File: pkg/utils/utils.go (hunk 7 above)
Comment (bn222) [reply pkg/utils/utils.go:803]: Actually, this is a good point. Let's also use the systemd package in go to start services

Comment (bn222) [reply pkg/utils/utils.go:803]: Actually, we have to cordon berfore and uncordon after like 803.

**Comment by SchSeba** [comment]
Hi @SalDaniele please rebase this PR.

also I have some questions here
1. is the hw configuration on ovs permanent or it needs to be done after every reboot?
2. does it needs to be done BEFORE you have workloads in the system?
3. this will only work if there is the ovs binary commands on the host

@Eoghan1232 be aware this change can impact the `/host` mount reduction you are working on

**Comment by SalDaniele** [comment]
> Hi @SalDaniele please rebase this PR.
> 
> also I have some questions here
> 
> 1. is the hw configuration on ovs permanent or it needs to be done after every reboot?

Whatever we configure ovs-hwol to be will persist after each reboot. IIUC a reboot or restart is needed to apply the configuration once we set it with `ovs-vsctl set Open_vSwitch ... hw-offload=true`

> 2. does it needs to be done BEFORE you have workloads in the system?

This is a good point I will need to look into this. When we set it with the systemd file I believe that would mean ovs would always be configured before the workloads. I was triggering a node reboot after setting ovs-vsctl but this resulted in additional reboots when creating / removing sriov policies, so I opted to instead attempt to restart ovs. I wonder if cordoning before and after the restart as Balazs had mentioned would solve this.

> 3. this will only work if there is the ovs binary commands on the host
> 

Yes, I have some pending changes I will push w/ the rebase to put this behind a check `if utils.ClusterType == utils.ClusterTypeOpenshift`.

> @Eoghan1232 be aware this change can impact the `/host` mount reduction you are working on

**Comment by bn222** [comment]
regarding '2.' we need a cordone/uncordone before/after we restart ovs. This will only happen the first time we enable it. Once we have it enabled, the "hwol == true" will be persistent across reboots. Next time, when ovs starts up, it will already be with hwol enabled and the whole cordone/uncordone dance will be skipped.

The main benefit of this patch is that we get rid of another file we put on the system. This lgtm.

**Comment by SchSeba** [comment]
@SalDaniele can we close this one there is a huge work on enable switchdev in the sriov operator
